employee-rostering-frontend/src/store/alert/types.ts

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Action } from 'redux';
+import { AlertInfo } from 'store/alert/types';
+
+export enum ActionType {
+  ADD_ALERT = 'ADD_ALERT',
+  REMOVE_ALERT = 'REMOVE_ALERT'
+}
+
+export interface AddAlertAction extends Action<ActionType.ADD_ALERT> {
+  readonly alertInfo: AlertInfo;
+}
+
+export interface RemoveAlertAction extends Action<ActionType.REMOVE_ALERT> {
+  readonly id: number;
+}
+
+export type AlertAction = AddAlertAction | RemoveAlertAction;
+
+export interface AlertInfo {
+  id?: number;
+  createdAt?: Date;
+  i18nKey: string;
+  variant: "success" | "danger" | "warning" | "info";
+  params: any;
+  components?: React.ReactNode[];
```

Comment:
All the above is OK but the `components?: React.ReactNode[];` is something I strongly dislike based on what's *recommended* by [Redux documentation](https://redux.js.org/faq/organizing-state#can-i-put-functions-promises-or-other-non-serializable-items-in-my-store-state). Please take a look at it and tell me what you think. Are there any good reasons for this approach that I might be missing? Is there a hidden issue that might prevent the separation (store: only app state, ui: only React components)?

https://github.com/reduxjs/redux/issues/1793#issuecomment-224002596
> not that things like that are impossible. It's that they are non-idiomatic, and in the vast majority of use cases, not a good idea.

This is closest to my opinion. Based on what's recommended by the documentation and generally perceived as *idiomatic* I classify this (putting React nodes) as something we should avoid.

Nice article about Redux store structuring called [re-ducks](https://github.com/alexnm/re-ducks): https://www.freecodecamp.org/news/scaling-your-redux-app-with-ducks-6115955638be/.

Reply:
The main reason there is an optional `components` tab is that some alert messages have structure that need to be preserved. In our case, the "See Stack Trace" button which links to an unique Stack Trace Dialog. The alternatives are to use Functions (another no-no), put a map somewhere that map strings/numbers to functions/components (which seems like bad design), put react components as state (another no-no). The best alternative is probably the map which would be linked to a new component, but in turn it will have a higher maintenance cost.

Reply:
Well, you are putting a React component (`ServerSideExceptionDialog`) into the app state, aren't you? It's inside `showServerError()` method on line 74. That *is* no-no, do you agree?

I think the structure of a stack trace (lines plus caused-by hierarchy) is captured by `ServerSideExceptionInfo` so why not store just that in the state, without wrapping it into React code?

Reply:
The main issue is the "Link", which we need to tell the Alert Component is one way or another is a link instead of text. Currently testing out a solution that uses a map.